libp2p/autotls/acme.nim

```diff
@@ -0,0 +1,321 @@
+import options, json, base64, sequtils
+from times import DateTime, parse
+import chronos/apps/http/httpclient, jwt, results, bearssl/pem
+
+import ./utils
+import ../crypto/crypto
+import ../crypto/rsa
+import ../transports/tls/certificate_ffi
+import ../transports/tls/certificate
+
+const
+  LetsEncryptURL* = "https://acme-v02.api.letsencrypt.org"
+  LetsEncryptURLStaging* = "https://acme-staging-v02.api.letsencrypt.org"
+  Alg = "RS256"
+  DefaultChalCompletedRetries = 10
+  DefaultChalCompletedRetryTime = 1.seconds
+  DefaultFinalizeRetries = 10
+  DefaultFinalizeRetryTime = 1.seconds
+  DefaultRandStringSize = 256
+
+type ACMEAccount* = object
+  status*: Opt[string]
+  contact*: Opt[seq[string]]
+  key*: KeyPair
+  session*: HttpSessionRef
+  kid*: Opt[string]
+  newNonceURL: string
+  newAccountURL: string
+  newOrderURL: string
+  acmeServerURL: string
+
+proc new*(
+    T: typedesc[ACMEAccount],
+    key: KeyPair,
+    status: Opt[string] = Opt.none(string),
+    contact: Opt[seq[string]] = Opt.none(seq[string]),
+    kid: Opt[string] = Opt.none(string),
+    acmeServerURL: string = LetsEncryptURL,
+): Future[ref ACMEAccount] {.async: (raises: [ACMEError, CancelledError]).} =
+  let session = HttpSessionRef.new()
+  var directory: JsonNode
+  try:
+    let directoryResponse =
+      await HttpClientRequestRef.get(session, acmeServerURL & "/directory").get().send()
+    directory = bytesToString(await directoryResponse.getBodyBytes()).checkedParseJson()
+  except HttpError as exc:
+    raise newException(ACMEError, "Failed to connect to ACME server", exc)
+  except ValueError as exc:
+    raise newException(ACMEError, "Unable to parse JSON", exc)
+  except OSError as exc:
+    raise newException(ACMEError, "Unable to parse JSON", exc)
+  except IOError as exc:
+    raise newException(ACMEError, "Unable to parse JSON", exc)
+
+  let acc = new(ACMEAccount)
+  acc.status = status
+  acc.contact = contact
+  acc.kid = kid
+  acc.key = key
+  acc.session = session
+  acc.newNonceURL = directory.getJSONField("newNonce").getStr
+  acc.newAccountURL = directory.getJSONField("newAccount").getStr
+  acc.newOrderURL = directory.getJSONField("newOrder").getStr
+  acc.acmeServerURL = acmeServerURL
+  return acc
+
+proc newNonce(
+    self: ref ACMEAccount
+): Future[string] {.async: (raises: [ACMEError, CancelledError]).} =
+  try:
+    let resp =
+      await HttpClientRequestRef.get(self.session, self.newNonceURL).get().send()
+    return resp.headers.getString("Replay-Nonce")
+  except HttpError as exc:
+    raise newException(ACMEError, "Failed to request new nonce from ACME server", exc)
+
+# TODO: save n and e in account so we don't have to recalculate every time
+proc acmeHeader(
+    self: ref ACMEAccount, url: string, needsJwk: bool, kid: string = ""
+): Future[JsonNode] {.async: (raises: [ACMEError, CancelledError]).} =
```

Comment:
should return some user specified type, instead of `JsonNode`.

Reply:
This is a helper function to build an http header. The header will be sent as a json object. It makes no sense for the response of this function not to be json

Reply:
> The header will be sent as a json object.

work with type in code, real user defined types.  header should be be some user defined type with some fields `AcmeHeader`. 

the code that needs this header in json format will encode it in place where it needs it.